sound/soc/sof/ipc4-pcm.c

```diff
@@ -13,6 +13,46 @@
 #include "ipc4-priv.h"
 #include "ipc4-topology.h"
 
+static int sof_ipc4_set_multi_pipeline_state(struct snd_sof_dev *sdev, u32 state,
+					     struct snd_sof_pipeline_trigger_info *info)
+{
+	struct ipc4_pipeline_set_state_data data;
+	struct sof_ipc4_msg msg = {{ 0 }};
+	u32 primary, ipc_size;
+	int i;
+
+	primary = state;
+	primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_SET_PIPELINE_STATE);
+	primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+	msg.primary = primary;
+
+	msg.extension = SOF_IPC4_GLB_PIPE_MULTI;
+
+	data.count = 0;
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. So add the pipeline instance ID's starting with the
+	 * last one in the trigger list.
+	 */
+	for (i = info->count - 1; i >= 0; i--) {
+		struct snd_sof_widget *pipe_widget = info->pipelines[i];
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+		/* skip triggering backend widgets */
+		if (!pipeline->is_backend) {
+			data.pipeline_ids[i] = info->pipelines[i]->instance_id;
+			data.count++;
+		}
+	}
+
+	ipc_size = sizeof(u32) * (data.count + 1);
```

Comment:
It took me a while to figure out what that `+ 1` is doing there, hint: it is for the size of the `data.count`

Reply:
added a comment now